Add Url Unescape

troydai opened the pull request:
Issue: #860

dnfclas commented:
Hi __@troydai__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

troydai commented:
/cc @davidfowl

KrzysztofCwalina commented on src/System.Text.Encodings.Web.Utf8/UrlEncoder.cs:
```diff
@@ -1,19 +1,336 @@
+                    // The decodeReader iterator is always moved to the first byte not yet 
+                    // be scanned after the process. A failed decoding means the chars
+                    // between the reader and decodeReader can be copied to output untouched. 
+                    if (!DecodeCore(ref decodeReader, ref writer, end, data))
```
Why do you pass the "end" here? This information is already passed in data.

KrzysztofCwalina commented on src/System.Text.Encodings.Web.Utf8/UrlEncoder.cs:
```diff
@@ -1,19 +1,336 @@
         {
+            // preserves the original head. if the percent-encodings cannot be interpreted as sequence of UTF-8 octets,
+            // bytes from this till the last scanned one will be copied to the memory pointed by writer.
+            var byte1 = UnescapePercentEncoding(ref reader, end, buffer);
```
so this returns nullable byte. Do we have to do it? Can the calling code check the length of the buffer before calling this ? Would it be faster? I am asking because we need to be super careful with all these routines operating on single bytes. They can become quite expensive.

troydai replied:
i'll revert the nullable and instead return -1 for false cases.

troydai replied:
and the false case is not just about the length of the buffer but may also about the legitimacy of the byte representations follows the `%` therefore `UnescapsePercentEncoding` is responsible of determining if the next characters are acceptable.

KrzysztofCwalina commented on src/System.Text.Encodings.Web.Utf8/UrlEncoder.cs:
```diff
@@ -1,19 +1,336 @@
+                    return false;
+                }
+
+                var nextItr = reader;
```
Could you name it without using the shortening?

troydai replied:
will update.

KrzysztofCwalina commented on src/System.Text.Encodings.Web.Utf8/UrlEncoder.cs:
```diff
@@ -1,19 +1,336 @@
+            }
+
+            // all bytes are verified, write to the output
+            if (byteCount > 0)
```
It might be faster to write an int into the span. Otherwise you might be doing lots of range checks.

troydai replied:
could you elaborate here?

KrzysztofCwalina replied:
Feel free to postpone it (and add a TODO), but every time you write to the span, we do range checks today, and you are writing 4 times. Instead, you could create an int (with these 4 bytes) and use span.Write(int), which is in SpanExtensions to write one int.

troydai replied:
I got it. The tricky part here it does not always write 4 bytes. It may write any thing between 1 to 4 bytes. To fully the logic will be a bit complex:
```
IF 1 byte:
    Write 1 byte
ELSE-IF 2 bytes:
    Compose 1 short
    Write 1 short
ELSE-IF 3 bytes:
    Compose 1 short
    Write 1 short
    Write 1 byte
ELSE-IF 4 bytes:
    Compose 1 int
    Write 1 int
```
The code will be a myth to the maintainer. And I'm not sure about the performance gain given the additional comparison and value manipulation.

KrzysztofCwalina commented on src/System.Text.Encodings.Web.Utf8/UrlEncoder.cs:
```diff
@@ -1,19 +1,336 @@
+        /// <param name="scan">The value to read</param>
+        /// <param name="end">The end of the buffer</param>
+        /// <param name="buffer">The byte array</param>
+        /// <returns>The unescaped byte if success. Otherwise return -1.</returns>
```
What about null?

troydai replied:
updated. there won't be `null` return any more.

KrzysztofCwalina commented on src/System.Text.Encodings.Web.Utf8/UrlEncoder.cs:
```diff
@@ -1,19 +1,336 @@
+                throw new InvalidOperationException("The path contains null characters.");
+            }
+
+            if (byte1 <= 0x7F)
```
It seems like UnescapePercentEncoding can return -1. What happens then?

troydai replied:
updated. `-1` case will be return early.

KrzysztofCwalina commented on src/System.Text.Encodings.Web.Utf8/UrlEncoder.cs:
```diff
@@ -1,19 +1,336 @@
+        /// The unescape is done in place, which means the result span is the subset of 
+        /// the input span. The beginnings of the spans are the same.
+        /// </remarks>
+        public static int UnescapeInPlace(Span<byte> data)
```
I think we should call these Decode/Encode. It's consistent with other APIs we already shipped and with some terminology used even within this file.

KrzysztofCwalina replied:
I would just name the first parameter "source" and the only parameter "buffer". This is to be consistent with Base64 encoder we already have.

troydai replied:
ok. looks good to me. will update.

KrzysztofCwalina commented on tests/System.Text.Encodings.Web.Utf8.Tests/UnescapeUrlPathInPlaceTests.cs:
```diff
@@ -0,0 +1,18 @@
+            var len = UrlEncoder.UnescapeInPlace(input);
+            Assert.True(len <= input.Length);
+
+            var outputDecoded = Encoding.UTF8.GetString(input.Slice(0, len).ToArray());
```
You should use "new Utf8String(input.Slice(0, len).ToString()"

troydai replied:
will update.

troydai replied:
Skip for now, the outcome of `Utf8String.ToString` is inconsistent with `Encoding.UTF8.GetString`. It may be an issue in `Utf8String`.

troydai replied:
To be more specific, there are valid UTF8 coded byte span rejected by `Utf8String` under the reason "invalid characters".

For example: `%F0%A4%AD%A2` is decoded to `𤭢` in `Encoding.UTF8` but rejected in `Utf8String`.

KrzysztofCwalina commented on tests/System.Text.Encodings.Web.Utf8.Tests/UnescapeUrlPathTests.cs:
```diff
@@ -0,0 +1,34 @@
+namespace System.Text.Encodings.Web.Utf8.Tests
+{
+    public class UnescapeUrlPathTests : UrlEncoderTests
+    {
```
What's the reason for the base type?

troydai replied:
two sets of tests. once for unescapse in place and the other for unescape and copy to a new byte span. the scenarios are the same but the verification are different. in one it will assert that it indeed an in-place operation, and for the other one it ensure the data is copied to another array.

uses a base class here allows two sets of test cases to reuse the same set of scenarios while offer different test logic.

troydai replied:
yup. will add.

troydai commented on src/System.Text.Encodings.Web.Utf8/UrlEncoder.cs:
```diff
@@ -1,19 +1,336 @@
+        public static int UnescapeInPlace(Span<byte> data)
+        {
+            // the slot to read the input
+            var reader = 0;
```
sure.

troydai commented on src/System.Text.Encodings.Web.Utf8/UrlEncoder.cs:
```diff
@@ -1,19 +1,336 @@
+        /// </summary>
+        /// <param name="data">The byte span represents a UTF8 encoding url path.</param>
+        /// <returns>
+        /// The length of the unescape sequence on the input byte span. The sequence begins
```
will update.

troydai commented on src/System.Text.Encodings.Web.Utf8/UrlEncoder.cs:
```diff
@@ -1,19 +1,336 @@
+            return true;
+        }
+
+        private static void Copy(int begin, int end, ref int writer, Span<byte> buffer)
```
will do.

troydai replied:
does `Span.CopyTo` work for the cases that the destination and source are overlapped?

davidfowl replied:
The CoreCLR version does, but the corefxlab one doesn't https://github.com/dotnet/coreclr/blob/SpanOfT/src/mscorlib/src/System/Span.cs#L352.

troydai replied:
Will do.

troydai replied:
I'll leave this one behind since the `Span` we uses here won't support overlap.

troydai replied:
I can spend some time on expanding `corefxlab`'s `Span` and then get back to this.

troydai commented:
/to: @KrzysztofCwalina @davidfowl 
Addressed all the feedbacks except:
1. Span.CopyTo is not used because the current implementation cannot deal with overlap memory blocks.
2. Doesn't merge decoded bytes into larger primitive type before writing to avoid range check. I state my reason in the comments.
3. Doesn't use `Utf8String` in test. It rejects a few samples which are acceptable to `Encoding.UTF8`. It may be a bug to `Utf8String`.

KrzysztofCwalina commented:
@troydai, thanks. looks good for now. I will try to look into the Utf8String issue and CopyTo

KrzysztofCwalina commented:
@troydai, please resolve the merge conflict. Thanks.

troydai commented:
Rebase and squash the commits. Waiting for the CI validation.

troydai commented:
/cc @KrzysztofCwalina updated
